### ISSUE TITLE
Suggestions / Features - embedded

### ISSUE THREAD

**Issue by someburner** [sisong/HPatchLite#14]
First of all- amazing job and thank you for making this!

For context, I have been using `jojodiff` with JanPatch to create embedded binary OTA patch updates. I have ran a script across a large variety of firmware releases and on average, the `tuz` diff size is general 1/2 the size of the jojodiff. And even for very divergent binaries, the diff is still ~30KB less. And on top of that, it runs much faster.

The main reason I found this was to find a non-GPL diff library, since I want to port the diff generation code to golang without GPL restriction.

1. As far as I can tell, the other compression algos supported by hdiffpatch don't offer that much better compression (and proven by your comparison chart). I would suggest adding [jojodiff](https://jojodiff.sourceforge.net/) to the comparison chart.

2. It seems like tuz-2k - tuz-5k both generate a pretty good diff, beyond that the diffsize vs. ram tradeoff isn't very enticing. It seems like for embedded applications, `tuz` is best. On the one hand, it would ideal to be able to just add `hpatchlite` as a submodule and define implementation methods, but in the current repo structure you have to pull in sources from `hdiffpatch`, `hdiffpatchlite`, and `tinyuz`. In comparison to [janpatch](https://github.com/janjongboom/janpatch) this is a very tedious process for embedded developers. It would be nice to have 1-2 header files. This would probably help adoption, if that's important to you.

3. On smaller embedded devices running a bare-minimum OS, if any, POSIX apis are non-existent, overkill, or require special consideration. I got hdiffpatch working using something like this [bdfile.h](https://github.com/janjongboom/mbed-delta-update/blob/master/BDFile.h) wrapper that provides POSIX-style API for a regular SPI Flash chip. However, it's a little confusing in the example what is `_do_readOld` vs `_do_readFile` in the example program. For embedded, we may need to have different implementations of `_do_readFile`- e.g. for old file data we want to read the currently running program flash, and for patch/new file we want to read/write to external SPI flash chip. I have implemented this in `bdfile.h`, but it would be helpful to have more documentation on what each stub is used for.

4. Kinda similar to 3, but for embedded we often don't want to pull in anything from stdlib. Is `malloc` even necessary? There are `#ifdef`s for not using errno, etc, but it would be nice to have a single config file. Also, it would be nice to make it so that `assert` is compiled in but simply returns an error so the the patch method can safely exit. Although it seems the asserts are more for debugging and shouldn't trigger on bad inputs?

5. Would also be nice to have the option to not use filenames and instead have an enum for old/diff/new. For non-posix filenames and directories are also unnecessary. As well as open modes ("rb", "w+", etc).

6. Consider using a code formatter like [uncrustify](https://github.com/uncrustify/uncrustify). 

In summary, the code works well and didn't actually need many tweaks for non-POSIX, but with a little re-organizing it could be made a lot easier to integrate for embedded.

**Comment by sisong** [comment]
> "but with a little re-organizing it could be made a lot easier to integrate for embedded."    

I'm running HPatchLite+tinyuz patch on an MCU device, with 8 bit CPU, 4k RAM, no stdlib & no filesys.   
   
**HPatchLite** project is just hpatch_lite API's use demo, all diff&patch source code in HDiffPatch.   
**tuz**--tinyuz is a lossless compression algorithm, you can also choose other compression algorithms to use with hpatch_lite.    
   
If you need to run hpatch_lite on device, all code files in `HDiffPatch\libHDiffPatch\HPatchLite\`, and call `hpatch_lite_open()\hpatch_lite_patch()` yourself.   
If defined `NDEBUG` or `assert` macro, that func `assert()` is not required.   
There is no dependency filesys or any other header file, and that func `malloc()` is not required.   
   
if used tuz decompresser on the device, all code files in `tinyuz\decompress\`; recommended to define `_IS_USED_C_MEMCPY=0` macro so that func `memcpy()` is not required.

For how to use hpatch_lite+tuz, see demo code in func `int hpatchi(...)`.

**Comment by someburner** [comment]
@sisong Thanks for the response.

Like I said, I got it working! It just could have been a lot easier with a little more documentation (although admittedly I'm not the best with abstractions). Most of my points would probably not come up with a little documentation. I don't mind looking through the code (e.g. to see if `malloc` and `free` are used during the patch process), but it's not immediately clear without grepping all the sources.

I will probably just put up a stripped down embedded version to give people an idea how to do it with emulated POSIX calls. I understand the diff/patch code is separate from compressor used, and the reasons for doing so. Just throwing my 2 cents that an hpatchlite+tuz repo with only required sources would be much easier to integrate and maintain for embedded users, although obviously more work for you (or whoever would maintain that).

I understand `assert` is not *required*, but my question is: are there ever cases where given `old.bin` and `diff.bin`, if the data is bad/corrupted, if it might trigger an `assert` (if enabled), or are those all covered by the `_CHECK` macros?

Maybe another way to put it, has `hdiffpatch` / `hpatchlite` been fuzzed significantly?

**Comment by someburner** [comment]
> I'm running HPatchLite + tinyuz on an MCU device, with 8 bit CPU, 4k RAM, no stdlib & no filesys.

I think that would be a helpful demo if possible to put on github :)

**Comment by sisong** [comment]
> been fuzzed significantly?

yes, fuzze testing code at `HDiffPatch\test\unit_test.cpp`, when defined `_AttackPacth_ON`
   
> are there ever cases where given old.bin and diff.bin, if the data is bad/corrupted, if it might trigger an assert (if enabled), or are those all covered by the _CHECK macros?

`assert` is only used to check for error codes that shouldn't be there. `_CHECK` is not recommended to be modify, it check fail that are likely to occur, such as data read fail.       
defined macro `__RUN_MEM_SAFE_CHECK=1` default, will always checked `data is bad/corrupted`.  If the data is trusted, the safe check code can be removed by define macro `__RUN_MEM_SAFE_CHECK=0` yourself.

**Comment by someburner** [comment]
Okay thanks for clarifying. Looking forward to using this + tinyuz in more applications!
